---
title: "COD_20241127_MGK_IBS7048_week13_dendrogram_and_phylogenetc_trees"
author: "Minsik Kim"
date: "2024-11-27"
output:
        rmdformats::downcute:
        downcute_theme: "chaos"
codeolding: show
fig_width: 6
fig_height: 6
df_print: paged
editor_options: 
        chunk_output_type: inline
markdown: 
        wrap: 72
---
        
```{r warning=FALSE, message=FALSE, echo=FALSE, results='hide', setup}
#===============================================================================
#BTC.LineZero.Header.1.1.0
#===============================================================================
#R Markdown environment setup and reporting utility.
#===============================================================================
#RLB.Dependencies:
#   knitr, magrittr, pacman, rio, rmarkdown, rmdformats, tibble, yaml
#===============================================================================
#Input for document parameters, libraries, file paths, and options.
#=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=
knitr::opts_chunk$set(message=FALSE, warning = FALSE)

path_working <- 
         ifelse(sessionInfo()[1]$R.version$platform == "aarch64-apple-darwin20",
                "/Volumes/MacMini Drive/Dropbox (Personal)/Inha/5_Lectures/Advanced biostatistics/scripts/BTE3207_Advanced_Biostatistics/", 
                "/Users/minsikkim/Dropbox (Personal)/Inha/5_Lectures/Advanced biostatistics/scripts/BTE3207_Advanced_Biostatistics/")

path_library <- "/Library/Frameworks/R.framework/Resources/library/"
str_libraries <- c("pacman", "yaml", "knitr", "rmarkdown", "rmdformats")


YAML_header <-
        '---
title: "COD_20241127_MGK_IBS7048_week13_dendrogram_and_phylogenetc_trees"
author: "Minsik Kim"
date: "2024.11.27"
output:
    rmdformats::downcute:
        downcute_theme: "chaos"
        codeolding: hide
        fig_width: 6
        fig_height: 6
---'
seed <- "20241127"

#=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
#Loads libraries, file paths, and other document options.
#=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
FUN.LineZero.Boot <- function() {
        .libPaths(path_library)
        
        require(pacman)
        pacman::p_load(c("knitr", "rmarkdown", "rmdformats", "yaml","ggpubr"))
        
        knitr::opts_knit$set(root.dir = path_working)
        
        str_libraries |> unique() |> sort() -> str_libraries
        pacman::p_load(char = str_libraries)
        
        set.seed(seed)
}
FUN.LineZero.Boot()
#=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
#Outputs R environment report.
#=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
FUN.LineZero.Report <- function() {
        cat("Line Zero Environment:\n\n")
        paste("R:", pacman::p_version(), "\n") |> cat()
        cat("Libraries:\n")
        for (str_libraries in str_libraries) {
                paste(
                        "    ", str_libraries, ": ", pacman::p_version(package = str_libraries),
                        "\n", sep = ""
                ) |> cat()
        }
        paste("\nOperating System:", pacman::p_detectOS(), "\n") |> cat()
        paste("    Library Path:", path_library, "\n") |> cat()
        paste("    Working Path:", path_working, "\n") |> cat()
        paste("Seed:", seed, "\n\n") |> cat()
        cat("YAML Header:\n")
        cat(YAML_header)
}
FUN.LineZero.Report()


```


# Distrances of data

Trees can be generated from any kind of data.

First, we need to calculate distances 

data `USArrests` can be loaded using `data(USArrests)`, and this data set contains statistics, in arrests per 100,000 residents for assault, murder, and rape in each of the 50 US states in 1973. Also given is the percent of the population living in urban areas.



```{r}

library(tidyverse)
data(USArrests)

head(USArrests)
```

We can calculate the distanace of that data, using `dist()` function.

```{r}
US_dist <- dist(scale(USArrests), method = "euclidean")

US_dist
```

Based on the distance dataset, we can also calculate the *hierarchical* clusters of those distances (the levels of data)


```{r}

US_hc <- hclust(US_dist, method = "ward.D2")

US_hc

```

```{r}
US_hc %>% names
```

This hierarchical clustering data has `height` data, representing its level.

```{r}
US_hc$height %>% hist()


```

Using basic `plot()`function, we can visualize the distances and their hierarchical clustering groups.

```{r}
base::plot(US_hc)

plot(US_hc)
```

As the basic function is hard to manipulate (such as font sizes, font familiesm, etc.), we can use ggtree function from `ggtree` package

```{r}

library(BiocManager)
#BiocManager::install("ggtree")
library(ggtree)


ggtree(US_hc) + 
  geom_tiplab(size = 2, family = "serif")

```
Similarly, we can calculate distances between characters as well, using `stringdist` package

```{r}
#install.packages("stringdist")

library(stringdist)

```


stringdist package and `stringdist::stringdist` can calculate the distances between *character* values.

For examples,

```{r}

stringdist("a", "a")


stringdist(c("Inha", "Induck", "BiologicalEngineering"), 
           c("Inha", "Inha", "Inha"))


```

`stringdist::stringdistmatrix` generates the matrix of the distance calculations.

```{r}

stringdistmatrix(c("Inha", "Induck", "BiologicalEngineering"), 
           c("Inha", "Inha", "Inha"))


```

Using this feature, we can calculate the distrance and generate the hierarchical clustuing of *TEXTS*

```{r}
text <- c("Chiiiiken","CCCCChicken","Chickkkken","Cowwww","Cooow", "CCCCow")

distancemodels <- stringdist::stringdistmatrix(text, text)

distancemodels
```


```{R}
rownames(distancemodels) <- text

hc <- hclust(as.dist(distancemodels))

plot(hc)
```

Or

```{r}
ggtree(hc) + 
  geom_tiplab(size = 4, 
              family = "serif", 
              hjust = 0.5, vjust = 2)


```

As same as the above example, Genomic sequences can be compared as below.

```{r}

text <- c("AAAAAAAAAAAA",
          "AAAAAAAAAAAT", 
          "AAAAAAAAAAAC",
          "GAAAAAAAAAAA",
          "GGGGAAAAAAAA", 
          "GCGCGCGCGCGC")

distancemodels <- stringdist::stringdistmatrix(text, text)
rownames(distancemodels) <- text

hc <- hclust(as.dist(distancemodels))
plot(hc)

```
If we have a data.frame of those sequences (their taxa names), we can construct a phylogenetic tree!


```{r}


distancemodels
rownames(distancemodels) <- c("E.coli 1",
                              "E.coli 2",
                              "E.coli 3",
                              "E.albertii",
                              "S. blattae",
                              "H.Sapiens")
distancemodels
```

This data is having just the new label names, with the same distance matrix.

```{r}
hc <- hclust(as.dist(distancemodels))
plot(hc)

```
Else,

```{r}

ggtree(hc) + 
  geom_tiplab(size = 4, 
              family = "serif", 
              hjust = 0.5, vjust = 2)


```
To get hiearchical cluster levels, you can use `cutree` function.

```{r}


cutree(hc, k = 1:6) 

?cutree


```
You can rename the level values afterwards...

# Generating phylogenetic tree from sequence file

As same as above, you can generate phylogenetic tree for the sequence file.

You can download the example file at my github repo. *This is not mandatory, if you have the file already*

```{r}
#githubURL <- ("https://github.com/minsiksudo/IBS7048_Advanced_metagenomics/raw/21df48b7b9c7f80f48986ed3f583085cf3271cc1/phyloseq_example.rds")
#download.file(githubURL,"phyloseq_examples.rds") 
```


```{r}

library(phyloseq)

phyloseq <- readRDS("phyloseq_examples.rds")

refseq(phyloseq) %>% stringdistmatrix(.,.) 

```

# Constructing phylogenetic tree object

Actullay, the calculation for phylogenetic tree is not that simple.

We first construct a *neighbor-joining tree*, and then fit a GTR+G+I (Generalized time-reversible with Gamma rate variation) ***maximum likelihood tree*** using the neighbor-joining tree as a starting point.

find more information at 

https://en.wikipedia.org/wiki/Maximum_likelihood_estimation


*!!Warning!!*

This process may take longer than your expectation. Please check your computer is available for intensive computing. You may skip this line, as I uploaded the result data at my repo.


```{r}
#install.packages("phangorn")
library(phangorn)
#BiocManager::install("DECIPHER")
library(DECIPHER)

refseq(phyloseq)

```


```{r}
#aligning amplicons
alignment <- AlignSeqs(refseq(phyloseq), anchor=NA)

phang.align <- phyDat(as(alignment, "matrix"), type="DNA")

dm <- dist.ml(phang.align)

treeNJ <- NJ(dm) # Note, tip order != sequence order
fit = pml(treeNJ, data=phang.align)

## negative edges length changed to 0!

fitGTR <- update(fit, k=4, inv=0.2)

fitGTR <- optim.pml(fitGTR, model="GTR", optInv=TRUE, optGamma=TRUE,
                    rearrangement = "stochastic", 
                    control = pml.control(trace = 0))



```

Tree object is at 

```{r}
fitGTR$tree
```


# Merging the tree object into phyloseq tree

You can convert the data into phy_tree object

```{r}

phyloseq_tree <- merge_phyloseq(phyloseq, phy_tree(fitGTR$tree))

```


## Alternative!

I uploaded a phyloseq object that is already containiing the phylogenetic tree. It can be saved to hard disk using the below command.

```{r}
githubURL_tree <- ("https://github.com/minsiksudo/IBS7048_Advanced_metagenomics/raw/main/phyloseq_example_tree.rds")

download.file(githubURL_tree,"phyloseq_examples_tree.rds") 

phyloseq_tree <- readRDS("phyloseq_examples_tree.rds")

```


```{r}

phyloseq_tree

```

We can store this .tree file as a seperate file in our folder. To do that, we need `ape` package

```{r}

library(ape)

ape::write.tree(phyloseq::phy_tree(phyloseq_tree), "phy_tree.txt")

```



# Here we can see that this is a set of 63 phylogenetic trees. If we want to look
# at a specific gene tree, we can use the `[[]]` subsetting operator.


```{r}
tree_asv <- read.tree("phy_tree.txt")

```

```{r}
tree_asv
```


```{r}
ggtree(tree_asv) + 
  geom_tiplab(size = 2)


cutree(tree_asv, k = 1:10) 

```
Changing names of trees


```{r}

genus_names <- tax_table(phyloseq) %>% 
        data.frame %>%
        select(c("Genus"))


tree_asv_genus <- tree_asv

tree_asv_genus$tip.label <- genus_names[tree_asv_genus$tip.label, ]

ggtree(tree_asv_genus) + 
  geom_tiplab(size = 2)
```

The end.


# Bonus

How stringdist() function can be used for other fields?

```{r}

```




# Bibliography

```{r warning=FALSE, message=FALSE, echo=FALSE}
#===============================================================================
#BTC.LineZero.Footer.1.1.0
#===============================================================================
#R markdown citation generator.
#===============================================================================
#RLB.Dependencies:
#   magrittr, pacman, stringr
#=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
#BTC.Dependencies:
#   LineZero.Header
#===============================================================================
#Generates citations for each explicitly loaded library.
#=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
str_libraries <- c("r", str_libraries)
for (str_libraries in str_libraries) {
        str_libraries |>
                pacman::p_citation() |>
                print(bibtex = FALSE) |>
                capture.output() %>%
                .[-1:-3] %>% .[. != ""] |>
                stringr::str_squish() |>
                stringr::str_replace("_", "") |>
                cat()
        cat("\n")
}
#===============================================================================
```